### PR TITLE
network: generate network traffic by using iperf

### PR DESCRIPTION
**Pull request by WangXiangUSTC**
Signed-off-by: xiang <xiang13225080@163.com>

```shell
./bin/chaosd attack network flood -h
generate a mount of network traffic by using iperf

Usage:
  chaosd attack network flood [flags]

Flags:
      --duration string   number of seconds to run the iperf test (default "99999999")
  -h, --help              help for flood
  -i, --ip string         generate traffic to this IP address
      --parallel int32    number of iperf parallel client threads to run (default 1)
  -p, --port string       generate traffic to this port on the IP address
  -r, --rate string       the speed of network traffic, allows bps, kbps, mbps, gbps, tbps unit. bps means bytes per second

Global Flags:
      --log-level string   the log level of chaosd. The value can be 'debug', 'info', 'warn' and 'error'
      --uid string         the experiment ID
```

### CODE REVIEW

**Comment by cwen0** [comment]
@WangXiangUSTC `iperf ` tool wound not be installed on centos by default. Maybe we should add it to `tools` directory.

**Comment by cwen0** [comment]
```
2022-06-17T15:17:12.389+0800	INFO	background-process-manager	build command	{"command": "bash -c iperf -u -c  -t 99999999 -p 2333 -P 1 -b 1gkb"}
[2022/06/17 15:17:12.391 +08:00] [INFO] [network.go:349] ["Start iperf process successfully"] [command="/bin/bash -c iperf -u -c  -t 99999999 -p 2333 -P 1 -b 1gkb"] [Pid=12500]
2022-06-17T15:17:12.392+0800	INFO	background-process-manager	process stopped	{"pid": 12500}
Attack network successfully, uid: 26d7e208-5ae8-4eba-825d-345e18a9067b
```

This attack failed, but I can't find the error message from this log.

File: pkg/server/chaosd/network.go [hunk 1]
```diff
@@ -308,29 +308,49 @@ func (s *Server) applyEtcHosts(attack *core.NetworkCommand, uid string, env Envi
...
 	if err != nil {
-		return errors.WithStack(err)
+		return perrors.WithStack(err)
 	}
 	recoverFlag = false
 	return nil
 }
 
+func (s *Server) applyFlood(attack *core.NetworkCommand) error {
+	cmd := bpm.DefaultProcessBuilder("bash", "-c", fmt.Sprintf("iperf -u -c %s -t %s -p %s -P %d -b %s", attack.IPAddress, attack.Duration, attack.Port, attack.Parallel, attack.Rate)).
```
Comment (cwen0) [inline pkg/server/chaosd/network.go:335]: This command will run `iperf` as a client. Maybe we also need to support `iperf` as a serve and remind users the difference between `iperf` client and `iperf` serve

**Comment by cwen0** [comment]
@WangXiangUSTC PTAL

**Comment by WangXiangUSTC** [comment]
> ```
> 2022-06-17T15:17:12.389+0800	INFO	background-process-manager	build command	{"command": "bash -c iperf -u -c  -t 99999999 -p 2333 -P 1 -b 1gkb"}
> [2022/06/17 15:17:12.391 +08:00] [INFO] [network.go:349] ["Start iperf process successfully"] [command="/bin/bash -c iperf -u -c  -t 99999999 -p 2333 -P 1 -b 1gkb"] [Pid=12500]
> 2022-06-17T15:17:12.392+0800	INFO	background-process-manager	process stopped	{"pid": 12500}
> Attack network successfully, uid: 26d7e208-5ae8-4eba-825d-345e18a9067b
> ```
> 
> This attack failed, but I can't find the error message from this log.

Yeah, I forget to validate the `ip-address`, fixed in the last commit

File: pkg/server/chaosd/network.go (hunk 1 above)
Comment (WangXiangUSTC) [reply pkg/server/chaosd/network.go:335]: I updated the description of the command, only supporting the `client` mode now

**Comment by WangXiangUSTC** [comment]
> @WangXiangUSTC `iperf ` tool wound not be installed on centos by default. Maybe we should add it to `tools` directory.

I will add some description of install `iperf` in the document first, and maybe add this tool into `tools` directory in another pr.

**Review by cwen0 — APPROVED** [review]
LGTM

File: pkg/server/chaosd/network.go (hunk 1 above)
Comment (Andrewmatilde) [inline pkg/server/chaosd/network.go:335]: Why just UDP here?

**Comment by Andrewmatilde** [comment]
I find sometimes iperf cannot shutdown but force kill.
image

**Comment by cwen0** [comment]
> I find sometimes iperf cannot shutdown but force kill. image

@WangXiangUSTC PTAL

**Comment by Andrewmatilde** [comment]
This may be a Debian only problem.

File: pkg/server/chaosd/network.go (hunk 1 above)
Comment (WangXiangUSTC) [reply pkg/server/chaosd/network.go:335]: Because UDP provides unreliable transport, it can increase network traffic even without a server

**Comment by WangXiangUSTC** [comment]
> > I find sometimes iperf cannot shutdown but force kill. image
> 
> @WangXiangUSTC PTAL

I have already check it with @Andrewmatilde, it's a bug of `iperf` on some specified Linux distributed version. @cwen0

File: pkg/server/chaosd/network.go (hunk 1 above)
Comment (WangXiangUSTC) [reply pkg/server/chaosd/network.go:335]: It's more easier to use, users can generate network traffic without deploying a server.

**Review by Andrewmatilde — APPROVED** [review]
LGTM

**Pull request merged**